lib/process/child-pool.js

```diff
@@ -3,73 +3,57 @@
 var fork = require('child_process').fork;
 var path = require('path');
 var Promise = require('bluebird');
+var _ = require('lodash');
 
 module.exports = function ChildPool() {
   if(!(this instanceof ChildPool)){
     return new ChildPool();
   }
 
-  var retained = {};
-  var free = [];
+  this.retained = {};
+  this.free = [];
 
-  this.retain = function(processFile){
-    return new Promise(function(resolve, reject) {
+  this.retain = Promise.method(function(processFile){
+    var child = this.free.pop();
 
-      var child = free.pop();
-      if(child){
-        return resolve(child);
-      }
+    if (child) return child;
 
-      try{
-        var child = fork(path.join(__dirname, './master.js'));
+    child = fork(path.join(__dirname, './master.js'));
 
-        retained[child.id] = child;
+    this.retained[child.pid] = child;
 
-        child.on('exit', function(code, signal){
-          console.error('Child process exited', child.pid, code, signal);
+    child.on('exit', this.remove.bind(this, child));
```

Comment:
I think it would be nice to at least provide a debuglog when a child process as exited, since this could be valuable information fo the user when developing process functions.

Reply:
I agree with logging when developing bull, I but I strongly believe it should be disabled otherwise. The logging transport I use expects JSON, so this log breaks my logging setup. If you believe some developers will want this message, then the choice should be left to the developer via events IMO.